### PR TITLE
Add support for the creation of a nested stack containing the custom BYOS infrastructure

### PR DESCRIPTION
**Pull request by demartinofra**
* Log to stdout when PCLUSTER_LOG_TO_STDOUT env is set
This turn out to be useful when developing the CLI and quickly sending the logs to stdout rather than file. For example you can do: `PCLUSTER_LOG_TO_STDOUT=1 pcluster list-clusters`

* Add ClusterInfrastructure section to BYOS schema
Note no validation is implemented yet.
```
  ByosSettings:
    SchedulerDefinition:
      ByosVersion: "1.0"
      ClusterInfrastructure:
        CloudFormation:
          Template: https://url
      Events:
        HeadInitEnvironment:
          ExecuteCommand:
            Command: env
```

* Add scheduler_compute_resource_name to dna.json on compute nodes

* Add launch_templates_config.json on head node containing lt ids and version
File is written to `/opt/parallelcluster/shared/launch_templates_config.json`. Here is an example:
```
{
  "Queues": {
    "queue1": {
      "ComputeResources": {
        "compute-resource1": {
          "LaunchTemplate": {
            "Version": "1",
            "Id": "lt-0bde23e087c541ac1"
          }
        }
      }
    },
    "queue-spot": {
      "ComputeResources": {
        "compute-resource2": {
          "LaunchTemplate": {
            "Version": "1",
            "Id": "lt-04a0a6baf3ada9b94"
          }
        }
      }
    }
  }
}
```

* Add support for BYOS CloudFormation template
Support the creation of a BYOS substack when specified in config.

* [integ-tests] Fix byos test

* Define IAM roles in the main stack scope
This was needed to keep the CFN logical id constant and avoid having to change integ tests.

* [template] Fix issue when compute security group is passed in config

* [integ-tests] Update placement group CFN resource query

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/utils.py [hunk 1]
```diff
@@ -310,3 +310,10 @@ def load_json_dict(file_path):
         json_content = json.load(file)
 
     return json_content
+
+
+def get_attr(obj, attribute):
+    """Get nested object attribute and return None if attr does not exist."""
+    for attribute in attribute.split("."):
```
Comment (yuleiwan) [inline cli/src/pcluster/utils.py:317]: we can break the for loop if obj is None

File: cli/src/pcluster/utils.py [hunk 2]
```diff
@@ -310,3 +310,10 @@ def load_json_dict(file_path):
         json_content = json.load(file)
 
     return json_content
+
+
+def get_attr(obj, attribute):
```
Comment (yuleiwan) [inline cli/src/pcluster/utils.py:315]: Minor: I'd use `attributes` for input parameter to distinguish from for loop parameter

File: cli/tests/pcluster/example_configs/byos.required.yaml [hunk 3]
```diff
@@ -10,8 +10,14 @@ Scheduling:
   Scheduler: byos
   ByosSettings:
     SchedulerDefinition:
-      Key1: Value1
-      Key2: Value2
+      ByosVersion: "1.0"
+      ClusterInfrastructure:
```
Comment (yuleiwan) [inline cli/tests/pcluster/example_configs/byos.required.yaml:14]: ClusterInfrastructure is not required based on the design doc

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 4]
```diff
@@ -1107,10 +1110,57 @@ def make_resource(self, data, **kwargs):
...
+    def make_resource(self, data, **kwargs):
+        """Generate resource."""
+        return ByosCloudFormationInfrastructure(**data)
+
+
+class ByosClusterInfrastructureSchema(BaseSchema):
+    """Represent the schema for ClusterInfrastructure schema in a BYOS plugin."""
+
+    cloud_formation = fields.Nested(
+        CloudFormationClusterInfrastructureSchema, metadata={"update_policy": UpdatePolicy.UNSUPPORTED}
```
[the remark this answers is not in the document]
Comment (demartinofra) [reply cli/src/pcluster/schemas/cluster_schema.py:1128]: yep, fixed

**Pull request merged**